File: algo2/repo5/c_5/ac.py
#!/usr/bin/env python

class MachineAC:
    class State:
        def __init__ (self, id):
            self.id   = id
            self.next = {}

        def has_key(self, x):
            return x in self.next

    def __init__ (self, terms):
        self.state  = [ MachineAC.State(0) ]
        self.output = [[]]
        self._make_goto(terms)
        self._make_failure()

    def _make_goto(self, terms):
        for term in terms:
            cur = self.state[0]
            for x in term:
                if not cur.has_key(x):
                    new = MachineAC.State( len(self.state) )
                    cur.next[x] = new
                    self.state.append(new)
                    self.output.append([])
                cur = cur.next[x]
            s = cur.id
            self.output[s].append(term)

    def _make_failure(self):
        failure = [0] * len(self.state)
        print(failure)
        queue   = [ 0 ]
        while len(queue) > 0:
            s = queue.pop(0)
            for x in self.state[s].next.keys():
                next = self.g(s, x)
                print(s, x, next)
                if next is not None:
                    queue.append(next)
                if s != 0:
                    f = failure[s]
                    while self.g(f, x) is None:
                        f = failure[f]
                    failure[next] = self.g(f, x)
                    self.output[next].extend( self.output[failure[next]] )
                    print(failure)
        self.failure = failure
        print(self.output)

    def g(self, s, x):
        if x in self.state[s].next:
            return self.state[s].next[x].id
        else:
            if s == 0:
                return 0
            else:
                return

    def match (self, query):
        s = 0
        for i in range(len(query)):
            while self.g(s, query[i]) is None:
                s = self.failure[s]
            s = self.g(s, query[i])
            for x in self.output[s]:
                print ('%d,%d => %s' % (i - len(x) + 1, i, x))

ac = MachineAC(['a', 'dab', 'ab'])

print ('in: xbabcdex')
ac.match('xbabcdex')

print ('in: abc')
ac.match('abc')

File: algo2/repo5/c_5/ac.cpp
#include <unordered_map>
#include <vector>
#include <queue>
#include <iostream>
#include <fstream>
#include <string>
#include <sys/time.h>

typedef struct StateMachine
{
    private:
        const int _id;
    public:
        std::unordered_map<char, StateMachine*> next;
        StateMachine(int id) : _id(id){}
        bool has_key(const char &x){
            if(this->next.find(x)==this->next.end()) return false;
            return true;
        }
        void add_next(const char &x, StateMachine* ptr){ next[x] = ptr; }
        StateMachine* get_next(const char &x){ return next[x]; }
        int get_id(){ return _id; }
} _STATEMACHINE;

class MachineAC
{
    public:
        MachineAC(std::vector<std::string> &patterns){
            state.push_back(new StateMachine(0));
            make_goto(patterns);
            make_failure();
        }
        ~MachineAC(){ for(int i = 0; i < state.size(); i++) delete state[i]; }
        void match(std::string &text){
            StateMachine *s = state[0];
            for(int i = 0; i < text.size(); i++){
                while(go(s, text[i]) == 0) s = failure[s->get_id()];
                s = go(s, text[i]);
                for(std::string found : output[s])
                    std::cout << found << "\tfrom " << i-found.size()+1 << " to " << i << std::endl;;
            }
        }

    private:
        std::vector<StateMachine*> state;
        std::vector<StateMachine*> failure;
        std::unordered_map<StateMachine*, std::vector<std::string>> output;
        void make_goto(std::vector<std::string> &patterns){
            for(std::string pattern : patterns){
                StateMachine *current = state[0];
                for(char& p : pattern){
                    if(!current->has_key(p)){
                        StateMachine *next = new StateMachine(state.size());
                        current->add_next(p, next);
                        state.push_back(next);
                    }
                    current = current->get_next(p);
                }
                output[current].push_back(pattern);
            }
        }
        void make_failure(){
            failure.resize(this->state.size());
            std::fill(failure.begin(), failure.end(), this->state[0]);
            std::queue<StateMachine*> queue;
            StateMachine *state;
            StateMachine *next;
            StateMachine *f;
            queue.push(this->state[0]);
            while(queue.size()){
                state = queue.front();
                for(auto itr = state->next.begin(); itr != state->next.end(); itr++){
                    next = go(state, itr->first);
                    if(next != 0) queue.push(next);
                    if(state != this->state[0]){
                        f = failure[state->get_id()];
                        while(go(f, itr->first) == 0) f = failure[f->get_id()];
                        failure[next->get_id()] = go(f, itr->first);
                        output[next].insert(output[next].end(),
                            output[failure[next->get_id()]].begin(), output[failure[next->get_id()]].end());
                    }
                }
                queue.pop();
            }
        }
        StateMachine* go(StateMachine *state, const char &x){
            if(state->has_key(x)){
                return state->get_next(x);
            }else if(state->get_id() == 0){
                return state;
            }else{
                return 0;
            }
        }
};

int main(void)
{
    std::vector<std::string> patterns{"12345"};
    MachineAC machine(patterns);
    std::ifstream ifs("./pi.dat");
    std::string text;
    struct timeval stime, etime;
	double sec;

    if(ifs.fail()){
        std::cout << "Failed opening file" << std::endl;
        return -1;
    }
    getline(ifs, text);
	gettimeofday(&stime, NULL);

    machine.match(text);

    gettimeofday(&etime, NULL);
	sec = (etime.tv_sec - stime.tv_sec) + 
			(etime.tv_usec - stime.tv_usec) / 1000000.0;
	printf("Elapsed Time %.4f [sec] \n", sec);
}